spec/controllers/api/v3/logins_controller_spec.rb

```diff
@@ -39,53 +40,15 @@
       end
     end
 
-    describe 'request with valid phone number, password and otp, but otp is expired' do
-      let(:db_user) do
-        Timecop.freeze(Date.current - 3) { FactoryBot.create(:user, password: password) }
-      end
-
-      let(:request_params) do
-        { user:
-            { phone_number: db_user.phone_number,
-              password: password,
-              otp: db_user.otp } }
-      end
-
-      it 'responds with http status 401' do
-        post :login_user, params: request_params
-        expect(response.status).to eq(401)
-        expect(JSON(response.body)).to eq('errors' => { 'user' => [I18n.t('login.error_messages.expired_otp')] })
-      end
-    end
-
-    describe 'request with valid phone number and password but otp mismatches' do
-      let(:request_params) do
-        { user:
-            { phone_number: db_user.phone_number,
-              password: '1234',
-              otp: 'wrong otp' } }
-      end
-
-      it 'responds with http status 401' do
-        post :login_user, params: request_params
-        expect(response.status).to eq(401)
-        expect(JSON(response.body))
-          .to eq('errors' => {
-                   'user' => [I18n.t('login.error_messages.invalid_otp')]
-                 })
-      end
-    end
```

Comment:
Since we have a separate tested class for authentication behavior, should we stub the `Result` from that class in these specs?

Reply:
Good question. Normally I like to avoid mocks and stubs for most situations unless its really necessary - so things like external services or really tricky legacy code come to mind. For things like this where the controller is really just doing 'flow control', I still like having the tests exercise the real code path, just to make sure everything integrates together.

I've usually ended up regretting using mocks/stubs in situations like this to replace application dependencies, and can sleep better at night with the little bit of additional confidence that using the real objects gives me. 😁